Deep inspection topics (#919)
kubev2v/forklift-documentation: documentation/modules/proc_inspecting-vms-cli.adoc

// Module included in the following assemblies:
//
// * documentation/doc-Migrating_your_virtual_machines/assemblies/assembly_migrating-from-vmware.adoc

:_mod-docs-content-type: PROCEDURE
[id="proc_inspecting-vms-cli_{context}"]
= Inspecting VMs by using the command line interface

[role="_abstract"]
You can run Deep Inspection by creating Conversion custom resources from the command line. This method is useful for automation and batch operations where you need programmatic control over the inspection process.

[IMPORTANT]
====
Deep Inspection is a Technology Preview feature only. Technology Preview features
are not supported with Red Hat production service level agreements (SLAs) and
might not be functionally complete. Red Hat does not recommend using them
in production. These features provide early access to upcoming product
features, enabling customers to test functionality and provide feedback during
the development process.
====

> .Prerequisites

Reviewer: please add to prereqs: CBT (change block tracking) must be enabled

* {vmw} vSphere source provider configured in {project-short}.
* {vmw} Virtual Disk Development Kit (VDDK) configured for the source provider.
* Cluster resources to run inspection worker pods.
* `oc` CLI tool installed and access to the {ocp-short} cluster.
* Managed object reference (moRef) ID of the VM you want to inspect.

.Procedure

. Create a Conversion CR YAML file for the VM you want to inspect. For example, `vm-inspection.yaml`:
+

> // SME REVIEW: Is this Conversion CR YAML structure correct? Verify apiVersion, spec.source.id, spec.source.name, spec.provider fields.

Author: Please check the YAML example in lines 35 - 49.

Reviewer: ```
apiVersion: forklift.konveyor.io/v1beta1
kind: Conversion
metadata:
  name: deep-inspection-solenoci-rhel9-luks
  namespace: openshift-mtv
  labels:
    conversion-type: DeepInspection
    vmID: vm-2216
spec:
  connection:
    secret:
      name: v8-secret
      namespace: openshift-mtv
  diskEncryption:
    secret:
      name: luks-disk-secret
      namespace: openshift-mtv
    type: LUKS
  targetNamespace: openshift-mtv
  type: DeepInspection
  vddkImage: 'quay.io/.../vddk:8'
  vm:
    id: vm-2216
    name: solenoci-rhel9-luks
    type: vsphere
  xfsCompatibility: false
```

Where:
- `diskEncryption` is optional and only required when the VM disks are LUKS encrypted
- `connection.secret` is secret for source provider
- `xfsCompatibility` depends if VM uses XFSv4 or not (true for XFSv4, XFSv5 false)
- `vmID` is MoRef of the VM from vmware
- `vm.id/name/type` is required and is MoRef and name of the vmware VM
- `targetNamespace` is required and specifies where the DI pod will be created
- `labels` are required

Author: Updated, thanks @solenoci

{namespace}::
Specifies the namespace where {project-short} is installed. The default is +{namespace}+.

<vm_moref_id>::
Specifies the managed object reference (moRef) ID of the VM. For information about retrieving the moRef ID, see xref:proc_retrieving-vmware-moref_{context}[Retrieving the moRef].

<vm_name>::
Specifies the name of the VM as it appears in the {vmw} vSphere inventory.

<provider_name>::
Specifies the name of the {vmw} vSphere provider CR.
--

. Create the Conversion CR:
+
[source,terminal,subs="attributes+"]
----
$ oc create -f vm-inspection.yaml
----
+
{project-short} creates an inspection pod to perform the inspection.

. Optional: To inspect multiple VMs concurrently, create additional Conversion CRs. {project-short} processes up to 10 concurrent inspections. Additional inspections are queued until resources become available.

.Verification

. Monitor the inspection pod status:
+
[source,terminal,subs="attributes+"]
----
$ oc get pods -n {namespace} | grep <vm_name>-inspection
----
+
The pod status should progress from `Pending` to `Running` to `Completed`.

. Retrieve the inspection results from the Conversion CR:
+

> // SME REVIEW: Is the jsonpath correct for inspection results? Should it be .status.inspectionResult or something else?

Author: Please check the path in line 97: `$ oc get conversion <vm_name>-inspection -n {namespace} -o json | jq '.status.inspectionResult'`

Reviewer: correct, when inspection succeeds, the results are stored in the `status.inspectionResult`. When inspection fails the field is missing. Users can check the status of inspection by checking the `status.phase` (Running, Pending, Successful, Failed, Canceled)

//
// * documentation/doc-Migrating_your_virtual_machines/assemblies/assembly_migrating-from-vmware.adoc

:_mod-docs-content-type: PROCEDURE
[id="proc_inspecting-vms-cli_{context}"]
= Inspecting VMs by using the command line interface

[role="_abstract"]
You can run Deep Inspection by creating Conversion custom resources from the command line. This method is useful for automation and batch operations where you need programmatic control over the inspection process.

[IMPORTANT]
====
Deep Inspection is a Technology Preview feature only. Technology Preview features
are not supported with Red Hat production service level agreements (SLAs) and
might not be functionally complete. Red Hat does not recommend using them
in production. These features provide early access to upcoming product
features, enabling customers to test functionality and provide feedback during
the development process.
====

.Prerequisites

* {vmw} vSphere source provider configured in {project-short}.
* {vmw} Virtual Disk Development Kit (VDDK) configured for the source provider.
* Changed Block Tracking (CBT) must be enabled.
* Cluster resources to run inspection worker pods.

> * `oc` CLI tool installed and access to the {ocp-short} cluster.

Reviewer 2: ```suggestion
* `oc` CLI tool installed
* Access to the {ocp-short} cluster
```
feels like should be two bulletpoints for pre-reqs

metadata:
  name: <vm_name>-inspection
  namespace: {namespace}
  labels:
    conversion-type: DeepInspection
    vmID: <vm_moref_id>
spec:
  connection:
    secret:
      name: <provider_secret_name>
      namespace: {namespace}
  diskEncryption:
    secret:
      name: <luks_secret_name>
      namespace: {namespace}
    type: LUKS
  targetNamespace: {namespace}
  type: DeepInspection
  vddkImage: <vddk_image>
  vm:
    id: <vm_moref_id>
    name: <vm_name>
    type: vsphere
  xfsCompatibility: <xfs_compatibility>
----
+
where:
+
--
`<vm_name>-inspection`::
Specifies a unique name for the `Conversion` CR. A common convention is to use the VM name followed by `-inspection`.

`<namespace>`::
Specifies the namespace where {project-short} is installed. The default is +{namespace}+.

`labels`::
Required. Must include `conversion-type: DeepInspection` and `vmID` set to the VM's moRef ID.

`<vm_moref_id>`::
Specifies the managed object reference (moRef) ID of the VM. For information about retrieving the moRef ID, see xref:proc_retrieving-vmware-moref_{context}[Retrieving the moRef].

`<provider_secret_name>`::
Specifies the name of the secret for the source provider.

`diskEncryption`::

> Optional. Only required when the VM disks are LUKS encrypted.

Reviewer 2: ```suggestion
Optional. Required only when the VM disks are LUKS encrypted.
```

`<luks_secret_name>`::
Specifies the name of the secret that contains the LUKS encryption password.

`targetNamespace`::
Required. Specifies where the Deep Inspection pod is created.

`<vddk_image>`::
Specifies the VDDK image to use. For example, `quay.io/kubev2v/vddk:8`.

`vm.id`, `vm.name`, `vm.type`::

> Required. Specifies the moRef ID and name of the VMware VM.

Reviewer 2: ```suggestion
Required. Specifies the moRef ID, name, and type of the {vmw} VM.
```

`<vm_name>`::
Specifies the name of the VM as it appears in the {vmw} vSphere inventory.

`<xfs_compatibility>`::
Specifies whether to use XFS compatibility mode. Set to `true` for XFSv4, `false` for XFSv5.
--

. Create the Conversion CR:
+
[source,terminal,subs="attributes+"]
----
$ oc create -f vm-inspection.yaml
----
+

> {project-short} creates an inspection pod to perform the inspection.

Reviewer 2: ```suggestion
{project-short} creates an inspection pod.
```

feels a bit redundant to have the inspection pod performing an inspection

. Optional: To inspect multiple VMs concurrently, create additional Conversion CRs. {project-short} processes up to 10 concurrent inspections. Additional inspections are queued until resources become available.

.Verification

. Monitor the inspection pod status:
+
[source,terminal,subs="attributes+"]
----
$ oc get pods -n {namespace} | grep <vm_name>-inspection
----
+
The pod status should progress from `Pending` to `Running` to `Completed`.

. Check the inspection status:
+
[source,terminal,subs="attributes+"]
----
$ oc get conversion <vm_name>-inspection -n {namespace} -o json | jq '.status.phase'
----
+
The status can be `Running`, `Pending`, `Successful`, `Failed`, or `Canceled`.

. Retrieve the inspection results from the Conversion CR:
+
[source,terminal,subs="attributes+"]
----
$ oc get conversion <vm_name>-inspection -n {namespace} -o json | jq '.status.inspectionResult'
----
+
The inspection results include information about the guest operating system, file systems, architecture, and any detected issues. The `inspectionResult` field is only present when the inspection succeeds.

> . After the inspection completes, the inspection pod and associated persistent volume claims (PVCs) are automatically cleaned up by {project-short}. To troubleshoot a failed inspection, view the pod logs:

Reviewer 2: ```suggestion
. {project-short} automatically cleans up the inspection pod and associated persistent volume claims (PVCs) after the inspection completes. To troubleshoot a failed inspection, view the pod logs:
```

Author: . After the inspection completes, {project-short} automatically cleans up the inspection pod and associated persistent volume claims (PVCs). To troubleshoot a failed inspection, view the pod logs:
